### PR TITLE
Updates and improvements for node-sass v3

### LINKED ISSUE #22: Doesn't log to console when sass file is invalid
There is no callback for checking errors nor logs to console. So, it is a bit hard to identify problem.
Maybe it's because I use it without connect module. 

```javascript
// middleware.js
module.exports = function(options, cb){ // line 44

if (err) { cb(err); return next(err); } // line 115

// when using
app.use('/stylesheets', sassMiddleware({
  src: "./public/stylesheets/",
  debug: true,
  outputStyle: 'expanded'
}, function(err){
   console.log(err);
}));
```

### LINKED ISSUE #31: If an imported file is not found request never ends
If an imported file doesn't exist the site is not able to load the stylesheet and stays waiting indefinitely.
There are no notifications in console.

image

### PR DESCRIPTION
**Pull request by bumxu**
This monster contains:
- Use of strict mode. :cop: 
- Updated dependencies.
- Updates and some bug fix for node-sass v3.
- Minor changes in formatting and documentation.
 - Removed unused/unsupported code, etc.
 - Additional comments.
 - Documented some undocumented options.
- This should fix #22 with a half-decent error output. :see_no_evil: 
- This should fix #31 ending the css request with status 500.

It includes commit iMoritz@1ea16c1 by @iMoritz fork. :punch: 

It needs new tests. :construction:

:yum:

### CODE REVIEW

**Comment by iMoritz** [comment]
Looks very nice :+1:

File: middleware.js [hunk 1]
```diff
@@ -143,16 +152,9 @@ module.exports = function(options){
...
             data: str,
-            success: function(result){
-              done(null, result);
-            },
-            error: function(err) {
-              done(err);
-            },
-            includePaths: [ sassDir ].concat(options.include_paths || options.includePaths || []),
-            imagePath: options.image_path || options.imagePath,
+            includePaths: [sassDir].concat(options.include_paths || options.includePaths || []),
```
Comment (am11) [inline middleware.js]: `options.include_paths` is retired. Lets only keep `options.includePaths`.

File: middleware.js [hunk 2]
```diff
@@ -118,12 +127,12 @@ module.exports = function(options){
...
           if (!options.response) {
-              mkdirp(dirname(cssPath), 0700, function(err){
-                  if (err) return error(err);
-                  fs.writeFile(cssPath, data, 'utf8', function(err) {
-                      if (err) return error(err);
-                  });
+            mkdirp(dirname(cssPath), '0700', function(err){
+              if (err) return error(err);
+              fs.writeFile(cssPath, data, 'utf8', function(err) {
+                if (err) return error(err);
```
Comment (am11) [inline middleware.js]: Lets try to avoid K&R style in favour of proper bracketing:

```javascript
if (err) {
  return error(err);
}

fs.writeFile(cssPath, data, 'utf8', function(err) {
  if (err) {
    return error(err);
  }
```

The extra line between first if-block and `fs.writeFile ..` is not the result of OCD, it is just a preferred design! :wink:

File: middleware.js [hunk 3]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
+          sassDir  = dirname(sassPath);
 
       if (root) {
-        cssPath = join(root, dest, path.replace(new RegExp('^' + dest), ''));
+        cssPath  = join(root, dest, path.replace(new RegExp('^' + dest), ''));
         sassPath = join(root, src, path
             .replace(new RegExp('^' + dest), '')
             .replace(/\.css$/, '.scss'));
-        sassDir = dirname(sassPath);
+        sassDir  = dirname(sassPath);
```
Comment (am11) [inline middleware.js]: extra space? please revert the change on line 97.

File: middleware.js [hunk 4]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
+      next();
+    } else {
+      var cssPath  = join(dest, path),
           sassPath = join(src, path.replace(/\.css$/, '.scss')),
-          sassDir = dirname(sassPath);
+          sassDir  = dirname(sassPath);
 
       if (root) {
-        cssPath = join(root, dest, path.replace(new RegExp('^' + dest), ''));
+        cssPath  = join(root, dest, path.replace(new RegExp('^' + dest), ''));
```
Comment (am11) [inline middleware.js]: extra space? please revert the change on line 93.

File: middleware.js [hunk 5]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
-    if (/\.css$/.test(path)) {
-      var cssPath = join(dest, path),
+
+    if (!/\.css$/.test(path)) {
+      next();
+    } else {
+      var cssPath  = join(dest, path),
           sassPath = join(src, path.replace(/\.css$/, '.scss')),
-          sassDir = dirname(sassPath);
+          sassDir  = dirname(sassPath);
```
Comment (am11) [inline middleware.js]: extra space? please revert the change on line 90.

File: middleware.js [hunk 6]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
   var root = options.root || null;
 
+  // Force compilation everytime
+  var force = options.force || options.response;
+  // Enable debug output
+  var debug = options.debug;
+
   // Default compile callback
-  options.compile = options.compile || function(){
+  options.compile = options.compile || function() {
```
Comment (am11) [inline middleware.js]: :+1:

File: middleware.js [hunk 7]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
+ *     connect.static(__dirname + '/public')
+ *   );
  *
  * @param {Object} options
  * @return {Function}
  * @api public
  */
 
-module.exports = function(options){
+module.exports = function(options) {
```
Comment (am11) [inline middleware.js]: :+1:

File: middleware.js [hunk 8]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
- *      var server = connect.createServer(
- *          sass.middleware({
- *              src: __dirname
- *            , dest: __dirname + '/public'
- *          })
- *        , connect.static(__dirname + '/public')
- *      );
+ *   var server = connect.createServer(
+ *     sass.middleware({
+ *       src:  __dirname,
```
Comment (am11) [inline middleware.js]: Extra space *before* `__dirname`

File: middleware.js [hunk 9]
```diff
@@ -1,3 +1,5 @@
+"use strict";
+
```
Comment (am11) [inline middleware.js]: :+1:

File: middleware.js [hunk 10]
```diff
@@ -94,22 +102,23 @@ module.exports = function(options){
...
-          data = err.message + ' in ' + fileLineColumn;
+          var file = sassPath;
+          if (err.file && err.file != 'stdin')
+            file = err.file;
+
+          var fileLineColumn = file + ':' + err.line + ':' + err.column;
+          data = err.message.replace(/^ +/, '') + '\n\nin ' + fileLineColumn;
           if (debug) logError(data);
+
           if (options.error) options.error(err);
```
Comment (am11) [inline middleware.js]: Please remove this line and see the comment on line 60.

File: middleware.js [hunk 11]
```diff
@@ -28,65 +33,68 @@ var imports = {};
...
-  // Source dir required
-  var src = options.src;
-  if (!src) { throw new Error('sass.middleware() requires "src" directory'); }
-
-  // Default dest dir to source
+  // Source directory (required)
+  var src = options.src || function() {
+    throw new Error('sass.middleware() requires "src" directory.');
+  }();
+  // Destination directory (source by default)
```
Comment (am11) [inline middleware.js]: Back reference line 118.

Please add the following before line 60:

```javascript
var error = typeof options.error === 'function' ? options.error || function (err) {
  throw new Error(err);
}
```

**Comment by am11** [comment]
Thanks guys! :+1: 

This looks good to me modulo @iMoritz's comment and those minor comments I left.

Also, v3.0.1 will be made available soon (probably today), so would be nice to bump the version before merging this PR.

Last but not the least, going by https://david-dm.org/sass/node-sass-middleware#info=devDependencies&view=table, please bump the version of mocha and connect (2.2.4 and 3.3.5 respectively).

File: middleware.js (hunk 1 above)
Comment (bumxu) [inline middleware.js]: `options.output_style` would be kept?

File: middleware.js (hunk 3 above)
Comment (bumxu) [inline middleware.js]: I have been reading about it. :astonished:
I'm sorry for this, I had the habit of align all...

File: middleware.js (hunk 2 above)
Comment (bumxu) [inline middleware.js]: +1

File: middleware.js (hunk 1 above)
Comment (am11) [inline middleware.js]: Nope, just `options.outputStyle`..

**Comment by bumxu** [comment]
Same syntax (^3.0.1) or fixed?
Currently all three are using caret `^`, so in principle we have green light.

File: middleware.js (hunk 11 above)
Comment (bumxu) [inline middleware.js]: Replacing current error?
```
// Ignore ENOENT to fall through as 404
var error = function(err) {
  next('ENOENT' == err.code ? null : err);
}
```

**Comment by am11** [comment]
@bumxu, 3.0.1 is not released yet, but it will happen :soon: &trade;. 

Perhaps, I did not follow the question but `^` means it will update you to the most recent major version (the first number). ^1.2.3 will match any 1.x.x release including 1.3.0, but will hold off on 2.0.0.

By the way, did you mean to close this PR or was it closed unintentionally? You can always change commits while the PR is opened. It is generally considered good to not to bleed one feature across the PRs for minor changes. This to keep the comment history and tracking of code changes.

File: middleware.js (hunk 11 above)
Comment (am11) [inline middleware.js]: Sorry, it should be:

```javascript
var error = typeof options.error === 'function' ? options.error || function (err) {
  throw err;
}
```

because node-sass here returns `Error` object.

**Comment by iMoritz** [comment]
@bumxu so you will fix things and get it ready to merge so I can go on with .sass support?

**Comment by bumxu** [comment]
That's what I mean, ^3.0.0 in principle is correct for 3.0.1 and beyond, ^2.1.0 for 2.2.4, etc...

Sorry, I know I've taken some risks but I would keep the repository as clean as possible.
